File: modules/home/scripts/wall.cpp
#include <array>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <map>
#include <memory>
#include <ncurses.h>
#include <sstream>
#include <string>
#include <vector>
#include <algorithm>

namespace fs = std::filesystem;

// Paths
const std::string HOME = std::getenv("HOME");
const fs::path CACHE_DIR = fs::path(HOME) / ".cache" / "personal";
const fs::path CMD_FILE = CACHE_DIR / "wallcmd";

// RAII wrapper for ncurses
struct NcursesGuard {
    NcursesGuard() {
        initscr();
        noecho();
        cbreak();
        keypad(stdscr, TRUE);
        curs_set(0);
    }
    ~NcursesGuard() {
        endwin();
    }
};

// Run shell command and capture output
std::string exec(const std::string &cmd) {
    std::array<char, 128> buffer{};
    std::string result;
    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd.c_str(), "r"), pclose);
    if (!pipe) return "";

    while (fgets(buffer.data(), buffer.size(), pipe.get())) {
        result += buffer.data();
    }
    return result;
}

std::vector<std::string> read_monitors() {
    std::vector<std::string> monitors;
    std::istringstream iss(exec("hyprctl monitors all | awk '/Monitor/ {print $2}'"));
    std::string line;
    while (std::getline(iss, line)) {
        if (!line.empty()) monitors.push_back(std::move(line));
    }
    return monitors;
}

// List image files in directory
std::vector<fs::path> list_images(const fs::path &dir) {
    std::vector<fs::path> images;
    for (auto &p : fs::directory_iterator(dir)) {
        if (!p.is_regular_file()) continue;
        std::string ext = p.path().extension().string();
        std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);
        if (ext == ".png" || ext == ".jpg" || ext == ".jpeg" || ext == ".bmp" || ext == ".gif") {
            images.push_back(p.path());
        }
    }
    return images;
}

// Display menu with ncurses and return index
int menu_select(const std::vector<std::string> &options, const std::string &title) {
    int highlight = 0;
    int choice = -1;
    int c;

    while (true) {
        clear();
        mvprintw(0, 0, "%s", title.c_str());
        for (size_t i = 0; i < options.size(); ++i) {
            if ((int)i == highlight) attron(A_REVERSE);
            mvprintw(i + 2, 0, "%s", options[i].c_str());
            if ((int)i == highlight) attroff(A_REVERSE);
        }
        refresh();

        c = getch();
        switch (c) {
            case 'k': case 'K':
                highlight = (highlight - 1 + options.size()) % options.size();
                break;
            case 'j': case 'J':
                highlight = (highlight + 1) % options.size();
                break;
            case 10:
                choice = highlight;
                break;
            case 'q': case 'Q':
                return -1;
        }
        if (choice != -1) break;
    }
    return choice;
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <wallpaper_dir> [mode]\n";
        return 1;
    }

    fs::path target_dir = fs::absolute(argv[1]);
    fs::path edited_dir = target_dir / "edited";
    fs::create_directories(edited_dir);
    fs::create_directories(CACHE_DIR);

    int mode = (argc >= 3) ? std::stoi(argv[2]) : 0;

    // === Mode 1: restore previous session ===
    if (mode == 1) {
        if (fs::exists(CMD_FILE) && fs::file_size(CMD_FILE) > 0) {
            std::cout << "Restoring last wallpaper session...\n";
            fs::permissions(CMD_FILE, fs::perms::owner_exec, fs::perm_options::add);
            system(CMD_FILE.string().c_str());
        } else {
            system("dunstify 'Wallpaper' 'No previous wallpaper session found.\nPlease run wallSet' -u normal -t 5000");
        }
        return 0;
    }

    auto monitors = read_monitors();
    if (monitors.empty()) {
        std::cerr << "No monitors found.\n";
        return 1;
    }

    auto images = list_images(target_dir);
    if (images.empty()) {
        std::cerr << "No images found in " << target_dir << "\n";
        return 1;
    }

    static const std::vector<std::string> rotation_labels = {"0°", "90°", "180°", "270°"};
    std::map<std::string, std::string> monitor_cmds;

    NcursesGuard ncurses; // auto cleanup

    while (true) {
        int m_idx = menu_select(monitors, "Select monitor:");
        if (m_idx == -1) break;

        std::vector<std::string> image_names;
        image_names.reserve(images.size());
        for (auto &p : images) image_names.push_back(p.filename().string());

        int w_idx = menu_select(image_names, "Select wallpaper:");
        if (w_idx == -1) break;

        int r_idx = menu_select(rotation_labels, "Select rotation:");
        if (r_idx == -1) break;
        int rotate = r_idx * 90;

        fs::path wallpaper = fs::absolute(images[w_idx]);
        fs::path final_wall = wallpaper;

        if (rotate != 0) {
            final_wall = edited_dir / (wallpaper.stem().string() + "_rot" + std::to_string(rotate) + wallpaper.extension().string());
            std::string cmd = "magick \"" + wallpaper.string() + "\" -rotate " + std::to_string(rotate) + " \"" + final_wall.string() + "\"";
            system(cmd.c_str());
        }

        std::string swww_cmd = "swww img -o \"" + monitors[m_idx] + "\" \"" + final_wall.string() + "\"";
        system(swww_cmd.c_str());
        monitor_cmds[monitors[m_idx]] = swww_cmd;

        std::vector<std::string> options = {"Set another wallpaper", "Exit"};
        int exit_choice = menu_select(options, "Options:");
        if (exit_choice == -1 || exit_choice == 1) break;
    }

    // Save wallcmd
    std::ofstream ofs(CMD_FILE);
    ofs << "#!/usr/bin/env bash\n\n";
    for (auto &p : monitor_cmds) ofs << p.second << "\n";
    ofs.close();
    fs::permissions(CMD_FILE, fs::perms::owner_all | fs::perms::group_read | fs::perms::others_read);

    std::cout << "Session saved to " << CMD_FILE << "\n";
    return 0;
}

File: modules/home/scripts/wall.py
#!/usr/bin/env python3

import curses
import os
import subprocess
from pathlib import Path

HOME = Path.home()

CACHE_DIR = HOME / ".cache" / "personal"
CMD_FILE = CACHE_DIR / "wallcmd"


# ---------- Helpers ----------


def run_cmd(cmd):
    """Run a shell command and return stdout as string."""
    try:
        out = subprocess.check_output(cmd, shell=True, text=True)
        return out.strip()
    except subprocess.CalledProcessError:
        return ""


def read_monitors():
    out = run_cmd("hyprctl monitors all | awk '/Monitor/ {print $2}'")
    return [line for line in out.splitlines() if line]


def list_images(dir_path: Path):
    pics = []
    for p in dir_path.iterdir():
        if p.is_file() and p.suffix.lower() in [
            ".png",
            ".jpg",
            ".jpeg",
            ".bmp",
            ".gif",
        ]:
            pics.append(p)
    return pics


# ---------- Ncurses Menu ----------


def menu(stdscr, title, options):
    curses.curs_set(0)
    highlight = 0

    while True:
        stdscr.clear()
        stdscr.addstr(0, 0, title)

        for i, opt in enumerate(options):
            if i == highlight:
                stdscr.addstr(i + 2, 0, opt, curses.A_REVERSE)
            else:
                stdscr.addstr(i + 2, 0, opt)

        key = stdscr.getch()

        if key in (ord("j"), ord("J")):
            highlight = (highlight + 1) % len(options)
        elif key in (ord("k"), ord("K")):
            highlight = (highlight - 1) % len(options)
        elif key == 10:  # Enter
            return highlight
        elif key in (ord("q"), ord("Q")):
            return -1


# ---------- Main Logic ----------


def main():
    import sys

    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <image_or_dir> [mode]")
        return 1

    input_path = Path(sys.argv[1]).resolve()
    mode = int(sys.argv[2]) if len(sys.argv) >= 3 else 0

    CACHE_DIR.mkdir(parents=True, exist_ok=True)

    # ------------------------------------------------------------
    # MODE 3: path is **image file** → instantly set wallpaper
    # ------------------------------------------------------------
    if mode == 3:
        monitors = read_monitors()
        if not monitors:
            print("No monitors found.")
            return 1

        if not input_path.is_file():
            print("Mode 3 requires a direct image file path.")
            return 1

        selected = input_path
        print(f"Setting '{selected.name}' on all monitors...")

        monitor_cmds = {}

        for m in monitors:
            cmd = f'swww img -o "{m}" "{selected}" --transition-type grow  --transition-duration 1 --transition-step 30'
            os.system(cmd)
            monitor_cmds[m] = cmd

        # Save session
        with open(CMD_FILE, "w") as f:
            f.write("#!/usr/bin/env bash\n\n")
            for _, c in monitor_cmds.items():
                f.write(c + "\n")

        CMD_FILE.chmod(0o755)
        print("Session saved to", CMD_FILE)
        return 0

    # ------------------------------------------------------------
    # MODE 1: restore previous session
    # ------------------------------------------------------------
    if mode == 1:
        if CMD_FILE.exists() and CMD_FILE.stat().st_size > 0:
            print("Restoring last wallpaper session...")
            CMD_FILE.chmod(0o755)
            os.system(str(CMD_FILE))
        else:
            os.system(
                "dunstify 'Wallpaper' 'No previous wallpaper session found.\nRun wallSet first.' -u normal"
            )
        return 0

    # ------------------------------------------------------------

    # NORMAL MODE (directory + TUI selector)
    # ------------------------------------------------------------

    target_dir = input_path
    if not target_dir.is_dir():
        print("Error: input must be a directory unless using mode 3.")
        return 1

    edited_dir = target_dir / "edited"
    edited_dir.mkdir(parents=True, exist_ok=True)

    monitors = read_monitors()
    if not monitors:
        print("No monitors found.")
        return 1

    images = list_images(target_dir)
    if not images:
        print(f"No images found in {target_dir}")
        return 1

    rotation_labels = ["0°", "90°", "180°", "270°"]
    monitor_cmds = {}

    def curses_app(stdscr):
        while True:
            m_idx = menu(stdscr, "Select monitor:", monitors)
            if m_idx == -1:
                break

            img_names = [p.name for p in images]
            w_idx = menu(stdscr, "Select wallpaper:", img_names)
            if w_idx == -1:
                break

            r_idx = menu(stdscr, "Select rotation:", rotation_labels)
            if r_idx == -1:
                break

            rotate = r_idx * 90
            wallpaper = images[w_idx]
            final_wall = wallpaper

            if rotate != 0:
                final_wall = (
                    edited_dir / f"{wallpaper.stem}_rot{rotate}{wallpaper.suffix}"
                )
                subprocess.run(
                    f'magick "{wallpaper}" -rotate {rotate} "{final_wall}"', shell=True
                )

            cmd = f'swww img -o "{monitors[m_idx]}" "{final_wall}"'
            os.system(cmd)
            monitor_cmds[monitors[m_idx]] = cmd

            opt = menu(stdscr, "Options:", ["Set another", "Exit"])
            if opt == -1 or opt == 1:
                break

    curses.wrapper(curses_app)

    # Save session script
    with open(CMD_FILE, "w") as f:
        f.write("#!/usr/bin/env bash\n\n")
        for _, cmd in monitor_cmds.items():
            f.write(cmd + "\n")

    CMD_FILE.chmod(0o755)
    print(f"Session saved to {CMD_FILE}")

    return 0


if __name__ == "__main__":
    main()
